File: v4/rust/iprange-livedb/src/validation/structure.rs
//! Explicit structure dictionary, reverse index, and ownership validation.

use crate::contract::{StructureKind, ValueKind};
use crate::error::Result;
use crate::format::Generation;
use crate::mapping::ByteSource;
use crate::membership_view;
use crate::structured_value::codec::{self, HashKey};
use crate::structured_value::{self, NetworkEnrichmentV1Codec, PayloadCodec};

use super::bitmap::{self, Kind, WordCache};
use super::context::Context;
use super::membership_table::{CountResult, InsertResult, Slot};
use super::structure_table;
use super::tree::{self, CellLayout, Codec};
use super::{ValidationObject, ValidationReason, ValidationSink};

struct HashCodec<P>(std::marker::PhantomData<P>);

impl<P: PayloadCodec> Codec for HashCodec<P> {
    type Key = HashKey;

    const BRANCH_TYPE: u8 = codec::HASH_BRANCH_TYPE;
    const LEAF_TYPE: u8 = codec::HASH_LEAF_TYPE;
    const AUX: u32 = P::KIND as u32;
    const BRANCH_LAYOUT: CellLayout = CellLayout::Fixed(codec::HASH_BRANCH_SIZE);
    const LEAF_LAYOUT: CellLayout = CellLayout::Fixed(codec::HASH_KEY_SIZE);
    const BRANCH_INVALID: ValidationReason = ValidationReason::StructureHashInvalid;
    const LEAF_INVALID: ValidationReason = ValidationReason::StructureHashInvalid;

    fn branch_key<B: ByteSource>(cell: B) -> Option<Self::Key> {
        codec::decode_hash_branch(cell).ok().map(|(key, _)| key)
    }

    fn branch_child<B: ByteSource>(cell: B) -> Option<u32> {
        codec::decode_hash_branch(cell).ok().map(|(_, child)| child)
    }

    fn leaf_key<B: ByteSource>(cell: B) -> Option<Self::Key> {
        codec::decode_hash(cell).ok()
    }
}

pub(crate) fn validate<S: ValidationSink>(context: &mut Context<'_, S>) -> Result<()> {
    if context.meta.value_kind != ValueKind::Structured {
        return Ok(());
    }
    match context.meta.structure_kind() {
        Some(StructureKind::NetworkEnrichmentV1) => {
            validate_kind::<NetworkEnrichmentV1Codec, S>(context)
        }
        Some(StructureKind::None) | None => context.emit(
            ValidationReason::StructureInvalid,
            ValidationObject::StructureDictionary,
            None,
            None,
            None,
        ),
    }
}

fn validate_kind<P: PayloadCodec, S: ValidationSink>(context: &mut Context<'_, S>) -> Result<()> {
    let mut maximum_id = 0;
    let id_root = context.meta.structure_id_root;
    let ids =
        structure_table::walk::<P, S, _>(context, id_root, |context, page, expected_id, cell| {
            validate_record::<P, S, _>(context, page, expected_id, cell, &mut maximum_id)
        })?;
    if ids.records != context.meta.structure_entry_count {
        count_mismatch(context, ValidationObject::StructureDictionary)?;
    }

    let mut previous = None;
    let hash_root = context.meta.structure_hash_root;
    let hashes = tree::walk::<HashCodec<P>, S, _>(
        context,
        hash_root,
        ValidationObject::StructureReverseIndex,
        |context, page, cell| validate_hash::<P, S, _>(context, page, cell, &mut previous),
    )?;
    if hashes.records != context.meta.structure_entry_count {
        count_mismatch(context, ValidationObject::StructureReverseIndex)?;
    }

    let used = bitmap::validate(
        context,
        context.meta.structure_used_root,
        context.meta.structure_id_limit,
        Kind::Structure,
    )?;
    finish(context, ids.records, used, maximum_id)
}

fn validate_record<P: PayloadCodec, S: ValidationSink, B: ByteSource>(
    context: &mut Context<'_, S>,
    page_number: u32,
    expected_id: u64,
    cell: B,
    maximum_id: &mut u32,
) -> Result<()> {
    let Ok(record) = codec::decode_record::<P, _>(cell) else {
        payload_finding(context, Some(page_number))?;
        return Ok(());
    };
    *maximum_id = (*maximum_id).max(record.id);
    if u64::from(record.id) != expected_id || expected_id >= context.meta.structure_id_limit {
        structure_finding(context, Some(page_number))?;
    }
    if record.refcount == 0 {
        refcount_finding(context, Some(page_number))?;
    }
    if structured_value::payload_digest::<P>(&record.payload)? != record.digest {
        hash_finding(context, Some(page_number))?;
    }
    let membership_id = P::membership_id(&record.payload);
    if !matches!(
        context.define_structure(record.id, record.refcount, membership_id, record.digest)?,
        InsertResult::Inserted
    ) {
        structure_finding(context, Some(page_number))?;
    }
    if membership_id != 0 {
        match context.count_membership_owner(membership_id) {
            CountResult::Full => membership_finding(context, Some(page_number))?,
            CountResult::Cancelled => return Err(crate::error::Error::Cancelled),
            CountResult::Unavailable => membership_finding(context, Some(page_number))?,
            CountResult::Inserted | CountResult::Existing => {}
        }
        if membership_view::by_id(context.mapping, &context.meta, membership_id, None).is_err() {
            membership_finding(context, Some(page_number))?;
        }
    }
    Ok(())
}

fn payload_finding<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page: Option<u32>,
) -> Result<()> {
    context.emit(
        ValidationReason::StructurePayloadInvalid,
        ValidationObject::StructureDictionary,
        page,
        None,
        None,
    )
}

fn validate_hash<P: PayloadCodec, S: ValidationSink, B: ByteSource>(
    context: &mut Context<'_, S>,
    page_number: u32,
    cell: B,
    previous: &mut Option<HashKey>,
) -> Result<()> {
    let Ok(key) = codec::decode_hash(cell) else {
        return Ok(());
    };
    if u64::from(key.id) >= context.meta.structure_id_limit
        || !context.mark_structure_reverse(key.id, key.digest)?
    {
        reverse_finding(context, Some(page_number))?;
    }
    if previous.is_some_and(|prior| prior.digest == key.digest) {
        let prior = previous.expect("checked above");
        if equal_payloads::<P, S>(context, prior.id, key.id).unwrap_or(true) {
            hash_finding(context, Some(page_number))?;
        }
    }
    *previous = Some(key);
    Ok(())
}

fn equal_payloads<P: PayloadCodec, S: ValidationSink>(
    context: &Context<'_, S>,
    left: u32,
    right: u32,
) -> Result<bool> {
    let generation = Generation::new(context.mapping, context.meta);
    let left = structured_value::find::<P, _>(
        &generation,
        context.meta.structure_id_root,
        context.meta.structure_id_limit,
        left,
    )?;
    let right = structured_value::find::<P, _>(
        &generation,
        context.meta.structure_id_root,
        context.meta.structure_id_limit,
        right,
    )?;
    Ok(matches!((left, right), (Some(left), Some(right)) if left.payload == right.payload))
}

fn finish<S: ValidationSink>(
    context: &mut Context<'_, S>,
    records: u64,
    used_bits: u64,
    maximum_id: u32,
) -> Result<()> {
    let mut defined = 0;
    let mut used = WordCache::new(
        context.meta.structure_used_root,
        context.meta.structure_id_limit,
        Kind::Structure,
    );
    for index in 0..context.structure_slots()? {
        context.checkpoint()?;
        let Some(slot) = context.structure_slot(index)? else {
            continue;
        };
        defined += u64::from(slot.defined);
        validate_slot(context, slot, &mut used)?;
    }
    let expected_limit = if maximum_id == 0 {
        1
    } else {
        u64::from(maximum_id) + 1
    };
    if defined != records
        || defined != context.meta.structure_entry_count
        || used_bits != defined
        || context.meta.structure_id_limit != expected_limit
    {
        structure_finding(context, None)?;
    }
    Ok(())
}

fn validate_slot<S: ValidationSink>(
    context: &mut Context<'_, S>,
    slot: Slot,
    used: &mut WordCache,
) -> Result<()> {
    if !slot.defined || slot.stored_refcount == 0 || slot.stored_refcount != slot.range_count {
        refcount_finding(context, None)?;
    }
    if slot.defined && !slot.reverse_seen {
        reverse_finding(context, None)?;
    }
    if slot.defined {
        let word = used.word(context, slot.id / 64).unwrap_or(0);
        if word & (1u64 << (slot.id % 64)) == 0 {
            structure_finding(context, None)?;
        }
    }
    Ok(())
}

fn count_mismatch<S: ValidationSink>(
    context: &mut Context<'_, S>,
    object: ValidationObject,
) -> Result<()> {
    context.emit(ValidationReason::RootCountInvalid, object, None, None, None)
}

fn structure_finding<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page: Option<u32>,
) -> Result<()> {
    context.emit(
        ValidationReason::StructureInvalid,
        ValidationObject::StructureDictionary,
        page,
        None,
        None,
    )
}

fn hash_finding<S: ValidationSink>(context: &mut Context<'_, S>, page: Option<u32>) -> Result<()> {
    context.emit(
        ValidationReason::StructureHashInvalid,
        ValidationObject::StructureDictionary,
        page,
        None,
        None,
    )
}

fn reverse_finding<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page: Option<u32>,
) -> Result<()> {
    context.emit(
        ValidationReason::StructureReverseIndexInvalid,
        ValidationObject::StructureReverseIndex,
        page,
        None,
        None,
    )
}

fn refcount_finding<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page: Option<u32>,
) -> Result<()> {
    context.emit(
        ValidationReason::StructureRefcountInvalid,
        ValidationObject::StructureDictionary,
        page,
        None,
        None,
    )
}

fn membership_finding<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page: Option<u32>,
) -> Result<()> {
    context.emit(
        ValidationReason::StructureMembershipInvalid,
        ValidationObject::StructureDictionary,
        page,
        None,
        None,
    )
}

File: v4/go/internal/validation/structure.go
package validation

// Structure validation (Rust validation/structure.rs): the dense ID
// table walk over the structure dictionary, the reverse-index hash tree
// walk with the adjacent same-digest collision compare, the membership
// ownership count, the used-bitmap count, and the bounded slot finish
// (refcount, reverse, and used arms). The reverse tables come from the
// bounded value-kind tables charged in the context.

import (
	"github.com/firehol/iprange/v4/go/internal/bitmap"
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/reader"
	"github.com/firehol/iprange/v4/go/internal/tree"
)

// validateStructure runs the structure table validators (Rust
// structure::validate): only structured databases carry a dictionary;
// an absent or unknown structure kind is its own invalid class.
func validateStructure(ctx *context) error {
	if ctx.meta.ValueKind != format.ValueKindStructured {
		return nil
	}
	switch ctx.meta.StructureKind {
	case format.StructureKindNetworkEnrichmentV1:
		return validateStructureKind(ctx)
	default:
		return ctx.emit(ReasonStructureInvalid, ObjectStructureDictionary, nil, nil, nil)
	}
}

// validateStructureKind runs the walks of one supported structure kind
// (Rust validate_kind).
func validateStructureKind(ctx *context) error {
	maximumID := uint32(0)
	catalog := reader.NewImmutable(ctx.mapping, ctx.meta)
	idResult, err := walkStructureTable(ctx, ctx.meta.StructureIDRoot, func(ctx *context, pageNumber uint32, expectedID uint64, cell []byte) error {
		return validateStructureRecord(ctx, pageNumber, expectedID, cell, &maximumID, catalog)
	})
	if err != nil {
		return err
	}
	if idResult.records != ctx.meta.StructureEntryCount {
		if err := structureCountMismatch(ctx, ObjectStructureDictionary); err != nil {
			return err
		}
	}
	var previous format.StructureHashKey
	hasPrevious := false
	hashResult, err := walkTree(ctx, ctx.meta.StructureHashRoot, ObjectStructureReverseIndex, structureHashCodec(ctx.meta.StructureKind), func(ctx *context, pageNumber uint32, cell []byte) error {
		return validateStructureHash(ctx, pageNumber, cell, &previous, &hasPrevious, catalog)
	})
	if err != nil {
		return err
	}
	if hashResult.records != ctx.meta.StructureEntryCount {
		if err := structureCountMismatch(ctx, ObjectStructureReverseIndex); err != nil {
			return err
		}
	}
	used, err := validateBitmap(ctx, ctx.meta.StructureUsedRoot, ctx.meta.StructureIDLimit, bitmap.KindStructure)
	if err != nil {
		return err
	}
	return finishStructure(ctx, idResult.records, used, maximumID)
}

// structureHashCodec is the Codec of the reverse-index tree (Rust
// HashCodec: fixed 36-byte hash leaves, fixed 40-byte branch entries,
// the structure kind as aux, and the hash class for both levels).
func structureHashCodec(kind uint8) treeCodec {
	return treeCodec{
		branchType:    byte(format.PageTypeStructureHashBranch),
		leafType:      byte(format.PageTypeStructureHashLeaf),
		aux:           uint32(kind),
		branchLayout:  format.FixedLayout(format.StructureHashBranchSize),
		leafLayout:    format.FixedLayout(format.StructureHashKeySize),
		branchInvalid: ReasonStructureHashInvalid,
		leafInvalid:   ReasonStructureHashInvalid,
		branchKey: func(cell []byte) (tree.Key, bool) {
			key, _, err := format.DecodeStructureHashBranchFields(cell)
			if err != nil {
				return tree.Key{}, false
			}
			return structureHashTreeKey(key), true
		},
		branchChild: func(cell []byte) (uint32, bool) {
			_, child, err := format.DecodeStructureHashBranchFields(cell)
			if err != nil {
				return 0, false
			}
			return child, true
		},
		leafKey: func(cell []byte) (tree.Key, bool) {
			key, err := format.DecodeStructureHashKey(cell)
			if err != nil {
				return tree.Key{}, false
			}
			return structureHashTreeKey(key), true
		},
	}
}

// structureHashTreeKey builds the ordered key of one hash record (Rust
// HashKey Ord: digest bytes, then id; the little-endian id keeps that
// byte order).
func structureHashTreeKey(key format.StructureHashKey) tree.Key {
	var raw [format.StructureHashKeySize]byte
	copy(raw[:32], key.Digest[:])
	format.PutU32(raw[32:36], key.ID)
	return tree.RawKey(raw[:])
}

// validateStructureRecord checks one dense-table record (Rust
// validate_record): the envelope and payload proof, the id-limit and
// slot windows, the nonzero refcount, the payload digest, the single
// define, and the membership ownership count.
func validateStructureRecord(ctx *context, pageNumber uint32, expectedID uint64, cell []byte, maximumID *uint32, catalog *reader.ImmutableReader) error {
	record, err := format.DecodeStructureRecord(cell, expectedID)
	if err != nil {
		return structurePayloadFinding(ctx, &pageNumber)
	}
	if record.ID > *maximumID {
		*maximumID = record.ID
	}
	if uint64(record.ID) != expectedID || expectedID >= ctx.meta.StructureIDLimit {
		if err := structureFinding(ctx, &pageNumber); err != nil {
			return err
		}
	}
	if record.Refcount == 0 {
		if err := structureRefcountFinding(ctx, &pageNumber); err != nil {
			return err
		}
	}
	digest, err := format.StructurePayloadDigest(ctx.meta.StructureKind, record.Payload)
	if err != nil {
		return err
	}
	if digest != record.Digest {
		if err := structureHashFinding(ctx, &pageNumber); err != nil {
			return err
		}
	}
	membershipID := format.U32(record.Payload[24:28])
	result, err := ctx.defineStructure(record.ID, record.Refcount, membershipID, record.Digest)
	if err != nil {
		return err
	}
	if result != InsertInserted {
		if err := structureFinding(ctx, &pageNumber); err != nil {
			return err
		}
	}
	if membershipID != 0 {
		switch ctx.countMembershipOwner(membershipID) {
		case CountFull:
			if err := structureMembershipFinding(ctx, &pageNumber); err != nil {
				return err
			}
		case CountCancelled:
			return &format.Error{Code: format.CodeCancelled, Detail: "validation cancelled"}
		case CountUnavailable:
			if err := structureMembershipFinding(ctx, &pageNumber); err != nil {
				return err
			}
		case CountInserted, CountExisting:
		}
		if _, err := catalog.LookupMembershipID(membershipID); err != nil {
			if err := structureMembershipFinding(ctx, &pageNumber); err != nil {
				return err
			}
		}
	}
	return nil
}

// validateStructureHash checks one reverse-index leaf record (Rust
// validate_hash): the id-limit window, the reverse mark, and the
// adjacent same-digest collision compare.
func validateStructureHash(ctx *context, pageNumber uint32, cell []byte, previous *format.StructureHashKey, hasPrevious *bool, catalog *reader.ImmutableReader) error {
	key, err := format.DecodeStructureHashKey(cell)
	if err != nil {
		// The codec leaf key already reported the undecodable class.
		return nil
	}
	marked, err := ctx.markStructureReverse(key.ID, key.Digest)
	if err != nil {
		return err
	}
	if uint64(key.ID) >= ctx.meta.StructureIDLimit || !marked {
		if err := structureReverseFinding(ctx, &pageNumber); err != nil {
			return err
		}
	}
	if *hasPrevious && previous.Digest == key.Digest {
		equal, err := equalStructurePayloads(catalog, previous.ID, key.ID)
		if err != nil || equal {
			if err := structureHashFinding(ctx, &pageNumber); err != nil {
				return err
			}
		}
	}
	*previous = key
	*hasPrevious = true
	return nil
}

// equalStructurePayloads proves two records carry identical payloads
// (Rust equal_payloads: an absent record is unequal, any read defect is
// reported as equal by the caller's unwrap_or(true)).
func equalStructurePayloads(catalog *reader.ImmutableReader, left, right uint32) (bool, error) {
	leftView, leftFound, err := catalog.LookupStructureID(left)
	if err != nil {
		return false, err
	}
	rightView, rightFound, err := catalog.LookupStructureID(right)
	if err != nil {
		return false, err
	}
	if !leftFound || !rightFound {
		return false, nil
	}
	leftValue, err := leftView.Value()
	if err != nil {
		return false, err
	}
	rightValue, err := rightView.Value()
	if err != nil {
		return false, err
	}
	return leftValue == rightValue, nil
}

// finishStructure proves the dictionary totals (Rust finish: the slot
// defined count, the used-bitmap bits, and the exact id limit against
// the maximum observed id).
func finishStructure(ctx *context, dictionaryRecords, usedBits uint64, maximumID uint32) error {
	defined, err := validateStructureSlots(ctx)
	if err != nil {
		return err
	}
	expectedLimit := uint64(1)
	if maximumID != 0 {
		expectedLimit = uint64(maximumID) + 1
	}
	if defined != dictionaryRecords || defined != ctx.meta.StructureEntryCount ||
		usedBits != defined || ctx.meta.StructureIDLimit != expectedLimit {
		return structureFinding(ctx, nil)
	}
	return nil
}

// validateStructureSlots walks every table entry (Rust validate_slots).
func validateStructureSlots(ctx *context) (uint64, error) {
	defined := uint64(0)
	used := newBitmapWordCache(ctx.meta.StructureUsedRoot, ctx.meta.StructureIDLimit, bitmap.KindStructure)
	slots, err := ctx.structureSlots()
	if err != nil {
		return 0, err
	}
	for index := 0; index < slots; index++ {
		if err := ctx.checkpoint(); err != nil {
			return 0, err
		}
		slot, ok, err := ctx.structureSlot(index)
		if err != nil {
			return 0, err
		}
		if !ok {
			continue
		}
		if slot.Defined {
			defined++
		}
		if err := validateStructureSlot(ctx, slot, &used); err != nil {
			return 0, err
		}
	}
	return defined, nil
}

// validateStructureSlot checks one occupied table entry (Rust
// validate_slot: refcount, reverse mark, and used bit; every arm runs
// and reports its own class).
func validateStructureSlot(ctx *context, slot Slot, used *bitmapWordCache) error {
	if err := validateStructureSlotRefcount(ctx, slot); err != nil {
		return err
	}
	if err := validateStructureSlotReverse(ctx, slot); err != nil {
		return err
	}
	return validateStructureSlotUsed(ctx, slot, used)
}

func validateStructureSlotRefcount(ctx *context, slot Slot) error {
	if !slot.Defined || slot.StoredRefcnt == 0 || slot.StoredRefcnt != slot.RangeCount {
		return structureRefcountFinding(ctx, nil)
	}
	return nil
}

func validateStructureSlotReverse(ctx *context, slot Slot) error {
	if slot.Defined && !slot.ReverseSeen {
		return structureReverseFinding(ctx, nil)
	}
	return nil
}

func validateStructureSlotUsed(ctx *context, slot Slot, used *bitmapWordCache) error {
	if slot.Defined {
		word, err := used.word(ctx, slot.ID/64)
		if err != nil {
			word = 0
		}
		if word&(uint64(1)<<(slot.ID%64)) == 0 {
			return structureFinding(ctx, nil)
		}
	}
	return nil
}

func structureCountMismatch(ctx *context, object ValidationObject) error {
	return ctx.emit(ReasonRootCountInvalid, object, nil, nil, nil)
}

// structureFinding streams the invalid class of a dictionary record,
// slot, or totals proof (Rust structure_finding).
func structureFinding(ctx *context, page *uint32) error {
	return ctx.emit(ReasonStructureInvalid, ObjectStructureDictionary, page, nil, nil)
}

// structurePayloadFinding streams the payload class of an undecodable
// dictionary record (Rust payload_finding).
func structurePayloadFinding(ctx *context, page *uint32) error {
	return ctx.emit(ReasonStructurePayloadInvalid, ObjectStructureDictionary, page, nil, nil)
}

// structureHashFinding streams the hash class of a dictionary record
// whose stored digest disagrees with its payload, or of an adjacent
// same-digest pair with equal payloads (Rust hash_finding).
func structureHashFinding(ctx *context, page *uint32) error {
	return ctx.emit(ReasonStructureHashInvalid, ObjectStructureDictionary, page, nil, nil)
}

// structureReverseFinding streams the invalid class of the
// reverse-index object (Rust reverse_finding).
func structureReverseFinding(ctx *context, page *uint32) error {
	return ctx.emit(ReasonStructureReverseIndexInvalid, ObjectStructureReverseIndex, page, nil, nil)
}

// structureRefcountFinding streams the refcount class of a dictionary
// record or slot (Rust refcount_finding).
func structureRefcountFinding(ctx *context, page *uint32) error {
	return ctx.emit(ReasonStructureRefcountInvalid, ObjectStructureDictionary, page, nil, nil)
}

// structureMembershipFinding streams the membership class of a record
// whose payload names an absent or unprovable membership (Rust
// membership_finding).
func structureMembershipFinding(ctx *context, page *uint32) error {
	return ctx.emit(ReasonStructureMembershipInvalid, ObjectStructureDictionary, page, nil, nil)
}
